### PR TITLE
Fix sync_dist for tpus

### LINKED ISSUE #6922: Add logic for sync distributed for TPUs in Step Result log
## 🐛 Bug

Add logic for sync distributed for TPUs in Step Result log

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/step_result.py#L106

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #6922 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/step_result.py [hunk 1]
```diff
@@ -105,10 +105,11 @@ def log(
 
         # sync across workers when using distributed training
         sync_fn = sync_fn or sync_ddp_if_available
+
         if sync_dist and isinstance(value, (torch.Tensor, numbers.Number)):
             is_dist_initialized = torch.distributed.is_available() and torch.distributed.is_initialized()
             # TODO: Find a way to make the reduction only once, so we don't need to clone.
-            if is_dist_initialized and isinstance(value, torch.Tensor):
+            if (is_dist_initialized or tpu_distributed) and isinstance(value, torch.Tensor):
```
Comment (awaelchli) [inline pytorch_lightning/core/step_result.py:112]: should we do `plugin.is_distributed here`?

**Review by awaelchli — COMMENTED** [review]
nice!

File: pytorch_lightning/core/step_result.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/core/step_result.py:112]: I don't think we can access the plugin object here.

Comment (awaelchli) [reply pytorch_lightning/core/step_result.py:112]: true! nvm then. looks good!

File: pytorch_lightning/utilities/distributed.py [hunk 2]
```diff
@@ -294,19 +295,13 @@ def register_ddp_comm_hook(
         )
     """
     if not _TORCH_GREATER_EQUAL_1_8:
-        rank_zero_warn(
-            "Not registering DDP comm hook. "
-            "To use communication hooks, please use pytorch>=1.8.0."
-        )
+        rank_zero_warn("Not registering DDP comm hook. " "To use communication hooks, please use pytorch>=1.8.0.")
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/utilities/distributed.py]: Done!

File: pytorch_lightning/utilities/distributed.py [hunk 3]
```diff
@@ -294,19 +295,13 @@ def register_ddp_comm_hook(
...
         return
     if ddp_comm_hook is None:
         return
     if ddp_comm_wrapper is not None:
         if not _TORCH_GREATER_EQUAL_1_9:
-            rank_zero_warn(
-                "Not applying DDP comm wrapper. "
-                "To use communication wrapper, please use pytorch>=1.9.0."
-            )
+            rank_zero_warn("Not applying DDP comm wrapper. " "To use communication wrapper, please use pytorch>=1.9.0.")
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/utilities/distributed.py]: haha, I missed this. Thanks!

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/distributed.py [hunk 4]
```diff
@@ -318,3 +313,9 @@ def register_ddp_comm_hook(
         state=ddp_comm_state,
         hook=ddp_comm_hook,
     )
+
+
+def tpu_distributed():
```
Comment (tchaton) [inline pytorch_lightning/utilities/distributed.py:318]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/utilities/imports.py [hunk 5]
```diff
@@ -88,3 +88,7 @@ def _compare_version(package: str, op, version) -> bool:
 _TORCHTEXT_AVAILABLE = _module_available("torchtext")
 _TORCHVISION_AVAILABLE = _module_available('torchvision')
 _XLA_AVAILABLE = _module_available("torch_xla")
+
+from pytorch_lightning.utilities.xla_device import XLADeviceUtils  # noqa: E402
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:92]: Move the import to the top? Also note that the `_*_AVAILABLE` vars are sorted

File: CHANGELOG.md [hunk 6]
```diff
@@ -240,6 +240,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `AttributeError for `require_backward_grad_sync` when running manual optimization with sharded plugin ([#6915](https://github.com/PyTorchLightning/pytorch-lightning/pull/6915))
 
 
+- Fixed `sync_dist` for tpus ([#6950](https://github.com/PyTorchLightning/pytorch-lightning/pull/6950))
```
Comment (carmocca) [inline CHANGELOG.md:243]: This could use a better explanation. Or was it always broken for all situations?

Comment (kaushikb11) [reply CHANGELOG.md:243]: Always broken. We never had sync dist logic for distributed tpus training.

File: pytorch_lightning/utilities/imports.py (hunk 5 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/imports.py:92]: We couldn't do that. As `XLADeviceUtils` imports `_XLA_AVAILABLE`.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/distributed.py (hunk 4 above)
Comment (ananthsub) [inline pytorch_lightning/utilities/distributed.py:318]:
```suggestion
def tpu_distributed() -> bool:
```

**Review by ananthsub — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Pull request merged**